### ISSUE TITLE
No way to cancel running queries 

### ISSUE THREAD

**Issue by kevinburkeshyp** [balderdashy/sails-postgresql#176]
The `pg` library offers an interface for canceling queries:

```javascript
  var client = new Client(config);
  var qry = "select pg_sleep(4)";
  var query1 = client.query(qry);
  client.cancel(client, query1);
```

I would like the ability to cancel queries that I've issued. Something like:

```javascript
var longRunningQuery = 'select * from users';
var returned = false;
query = Users.find().exec(function(err, users) {
    console.log(users);
})
setTimeout(function() {
    if (!returned) {
        query.cancel();
    }
}, 10000);
```

I *believe* you may be able to get partway there by changing the code in `spawnConnection` like so:

```javascript
    // Current way
    // Grab a client instance from the client pool
    pg.connect(connectionConfig, function(err, client, done) {
      after(err, client, done);
    });

    // New way
    client = new pg.Client(connectionConfig);
    client.connect(function(err, done) {
        after(err, client, done);
    });
    return client;
```

That would get you a client object in user space.

The next tricky part is that you'd need to pass the `Query` object into any cancel requests. In theory, there's nothing stopping you from building a query object before a connection has been established, and it's a little frustrating that this information isn't available at query time.

It could be nice to split out creating the client and creating the query from creating a new database connection and issuing the query. The result would be to reduce the amount of work done inside the `spawnConnection` callback - maybe you'd have a sync function called `getClient` which would retrieve configuration and a Client object and another function called `doConnect` or something which would apply transformations, connect to the database and hit the `logic` callback.

Failing that, you might be able to reproduce the same Query object, since it seems like creating a Query object isn't that complicated:

https://github.com/brianc/node-postgres/blob/master/lib/client.js#L314
https://github.com/brianc/node-postgres/blob/master/lib/client.js#L220

Replicating the Query object in user-land would be difficult because of the transformations applied by sails-postgresql to the query before it gets sent to `pg`. 

Thoughts?

**Comment by particlebanana** [comment]
This isn't something I've come up against but if node-postgres exposes a `cancel` method or the equivalent, I'm sure we can work out a way to return it and add it to the deferred object.

**Comment by kevinburkeshyp** [comment]
> if node-postgres exposes a cancel method or the equivalent,

It's a little tricky - you can't issue `cancel` on the `pg.Client` object because each client only has one query in progress at a time, so it would block until the query completes. The signature is `pg.cancel(config, client, query)`, where `client` is a `pg.Client` and `query` is the value returned synchronously by `client.query`. You can use `client.connectionParameters` for `config`.

https://github.com/brianc/node-postgres/wiki/pg#cancelconfig-client-query

https://github.com/brianc/node-postgres/blob/master/lib/index.js#L66

Here's the cancel code. Note it's on the Client object but you have to create a new one, you can't do it on the same client - might have got this wrong up above.

https://github.com/brianc/node-postgres/blob/master/lib/client.js#L219

**Comment by sailsbot** [comment]
Thanks for posting, @kevinburkeshyp.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
